crates/cxx-qt-build/src/lib.rs

```diff
@@ -888,55 +886,92 @@ impl CxxQtBuilder {
-            self.initializers
-                .push(include_str!("std_types_qt5.cpp").to_owned());
+            // If we are using Qt 5 then write the std_types source
+            // This registers std numbers as a type for use in QML
+            //
+            // Note that we need this to be compiled into an object file
+            // as they are stored in statics in the source.
+            //
+            // See also:
+            // https://github.com/rust-lang/rust/issues/108081
+            // https://github.com/KDAB/cxx-qt/pull/598
+            qt5_initializer = Some(include_str!("std_types_qt5.cpp").to_owned());
         }
-    }
 
-    fn generate_init_code(&self, initializers: &HashSet<PathBuf>) -> String {
-        initializers
+        let (declarations, calls): (Vec<_>, Vec<_>) = dependency_initializers
             .iter()
-            .map(|path| std::fs::read_to_string(path).expect("Could not read initializer file!"))
-            .chain(self.initializers.iter().cloned())
+            .map(|init_fun| {
+                (
+                    // declaration
+                    format!("extern \"C\" int {init_fun}();"),
+                    // call
+                    format!("{init_fun}();"),
+                )
+            })
+            .unzip();
+        let init_function = format!(
+            "{declarations}\nextern \"C\" int {init_fun}() {{\n{calls}\nreturn 42;\n}}",
+            init_fun = self.init_function_name(),
+            declarations = declarations.join("\n"),
+            calls = calls.join("\n"),
+        );
+
+        initializer_files
+            .iter()
+            .map(|file| std::fs::read_to_string(file).expect("Could not read initializer file"))
+            .chain(qt5_initializer)
+            .chain(Some(init_function))
             .collect::<Vec<_>>()
             .join("\n")
     }
 
-    fn build_initializers(&mut self, init_builder: &cc::Build, initializers: &HashSet<PathBuf>) {
+    fn build_initializers(&mut self, init_builder: &cc::Build, initializer_code: String) {
         let initializers_path = dir::out().join("cxx-qt-build").join("initializers");
         std::fs::create_dir_all(&initializers_path).expect("Failed to create initializers path!");
 
-        let initializers_path = initializers_path.join(format!("{}.cpp", crate_name()));
-        std::fs::write(&initializers_path, self.generate_init_code(initializers))
+        let initializers_file = initializers_path.join(format!("{}-init-lib.cpp", crate_name()));
+        std::fs::write(&initializers_file, initializer_code)
             .expect("Could not write initializers file");
+
+        // Build static initializers into a library that we link with whole-archive
+        // Then use object files just to trigger the initial call.
+        init_builder
+            .clone()
+            .file(initializers_file)
+            .link_lib_modifier("+whole-archive")
+            .compile(&format!("{}-initializers", crate_name()));
```

Comment:
is the whole-archive for the cargo build or cmake build or both ?

and then why is this needed ? can these initialiser functions not be public exported functions that are built into a library. Then when linking the library the object file that is also links makes use of the public functions preventing them being optimised out ?